File: backend-comic/src/main/java/com/thjvjpxx/backend_comic/service/CommentService.java
package com.thjvjpxx.backend_comic.service;

import java.util.List;

import com.thjvjpxx.backend_comic.dto.request.CommentRequest;
import com.thjvjpxx.backend_comic.dto.response.BaseResponse;
import com.thjvjpxx.backend_comic.dto.response.CommentResponse;
import com.thjvjpxx.backend_comic.enums.CommentStatus;

/**
 * Service xử lý API quản lý comment
 */
public interface CommentService {

    /**
     * Lấy tất cả comment với phân trang và tìm kiếm
     * 
     * @param page      Trang hiện tại
     * @param limit     Số lượng mỗi trang
     * @param search    Tìm kiếm theo tên
     * @param comicId   ID comic
     * @param chapterId ID chapter
     * @param userId    ID user
     * @param status    Trạng thái
     * @return Response chứa danh sách comment
     */
    BaseResponse<List<CommentResponse>> getAllComments(int page, int limit, String search,
            String comicId, String chapterId,
            String userId, CommentStatus status);

    /**
     * Lấy comment theo chapter với phân trang
     * 
     * @param chapterId ID chapter
     * @param page      Trang hiện tại
     * @param limit     Số lượng mỗi trang
     * @return Response chứa danh sách comment
     */
    BaseResponse<List<CommentResponse>> getCommentsByChapter(String chapterId, int page, int limit);

    /**
     * Lấy comment cha (parent comments) theo comic
     * 
     * @param comicId ID comic
     * @param page    Trang hiện tại
     * @param limit   Số lượng mỗi trang
     * @return Response chứa danh sách comment cha
     */
    BaseResponse<List<CommentResponse>> getParentCommentsByComic(String comicId, int page, int limit);

    /**
     * Lấy reply comments theo parent comment id
     * 
     * @param parentId ID parent comment
     * @return Response chứa danh sách reply comment
     */
    BaseResponse<List<CommentResponse>> getRepliesByParentId(String parentId);

    /**
     * Lấy chi tiết comment theo id
     * 
     * @param id ID comment
     * @return Response chứa comment
     */
    BaseResponse<CommentResponse> getCommentById(String id);

    /**
     * Tạo comment mới
     * 
     * @param request DTO chứa thông tin comment
     * @param userId  ID user
     * @return Response chứa comment đã tạo
     */
    BaseResponse<CommentResponse> createComment(CommentRequest request, String userId);

    /**
     * Xóa comment (soft delete)
     * 
     * @param id     ID comment
     * @param userId ID user
     * @return Response chứa comment đã xóa
     */
    BaseResponse<String> deleteComment(String id, String userId);

    /**
     * Chặn comment (chuyển status thành BLOCKED) - chỉ admin
     * 
     * @param id ID comment
     * @return Response chứa comment đã chặn
     */
    BaseResponse<String> blockComment(String id);

    /**
     * Bỏ chặn comment (chuyển status thành ACTIVE) - chỉ admin
     * 
     * @param id ID comment
     * @return Response chứa comment đã bỏ chặn
     */
    BaseResponse<String> unblockComment(String id);

    /**
     * Đếm số comment theo comic
     * 
     * @param comicId ID comic
     * @return Response chứa số lượng comment
     */
    BaseResponse<Long> countCommentsByComic(String comicId);

    /**
     * Đếm số comment theo chapter
     * 
     * @param chapterId ID chapter
     * @return Response chứa số lượng comment
     */
    BaseResponse<Long> countCommentsByChapter(String chapterId);
}

File: frontend-comic/src/services/commentService.ts
import { fetchApi } from "./api";
import { CommentResponse, CommentRequest, CommentFilters } from "@/types/comment";

const endpoint = '/comments';

// Lấy danh sách comment với phân trang và bộ lọc (cho admin)
export const getComments = async (
  page: number = 1,
  limit: number = 10,
  filters?: CommentFilters
) => {
  let fullEndpoint = `${endpoint}?page=${page - 1}&limit=${limit}`;

  if (filters?.search) {
    fullEndpoint += `&search=${encodeURIComponent(filters.search)}`;
  }

  if (filters?.comicId) {
    fullEndpoint += `&comicId=${filters.comicId}`;
  }

  if (filters?.chapterId) {
    fullEndpoint += `&chapterId=${filters.chapterId}`;
  }

  if (filters?.userId) {
    fullEndpoint += `&userId=${filters.userId}`;
  }

  if (filters?.status) {
    fullEndpoint += `&status=${filters.status}`;
  }

  return await fetchApi<CommentResponse[]>(fullEndpoint);
};

// Lấy chi tiết comment theo ID
export const getComment = async (id: string) => {
  return await fetchApi<CommentResponse>(`${endpoint}/${id}`);
};

// Tạo comment mới
export const createComment = async (data: CommentRequest) => {
  return await fetchApi<CommentResponse>(endpoint, {
    method: "POST",
    data: data,
  });
};

// Xóa comment (soft delete)
export const deleteComment = async (id: string) => {
  return await fetchApi<string>(`${endpoint}/${id}`, {
    method: "DELETE",
  });
};

// Chặn comment (admin only)
export const blockComment = async (id: string) => {
  return await fetchApi<string>(`${endpoint}/${id}/block`, {
    method: "POST",
  });
};

// Bỏ chặn comment (admin only)
export const unblockComment = async (id: string) => {
  return await fetchApi<string>(`${endpoint}/${id}/unblock`, {
    method: "POST",
  });
};

// Lấy comment theo chapter
export const getCommentsByChapter = async (
  chapterId: string,
  page: number = 1,
  limit: number = 10
) => {
  return await fetchApi<CommentResponse[]>(
    `${endpoint}/chapter/${chapterId}?page=${page - 1}&limit=${limit}`
  );
};

// Đếm comment theo comic
export const countCommentsByComic = async (comicId: string) => {
  return await fetchApi<number>(`${endpoint}/comic/${comicId}/count`);
};

// Đếm comment theo chapter
export const countCommentsByChapter = async (chapterId: string) => {
  return await fetchApi<number>(`${endpoint}/chapter/${chapterId}/count`);
};

// Lấy parent comments theo comic (bao gồm replies)
export const getParentCommentsByComic = async (
  comicId: string,
  page: number = 1,
  limit: number = 10
) => {
  return await fetchApi<CommentResponse[]>(
    `${endpoint}/comic/${comicId}/parents?page=${page - 1}&limit=${limit}`
  );
};

// Lấy replies theo parent comment ID
export const getRepliesByParentId = async (parentId: string) => {
  return await fetchApi<CommentResponse[]>(`${endpoint}/${parentId}/replies`);
};